sympy/solvers/recurr.py

```diff
@@ -727,20 +727,19 @@ def rsolve(f, y, init=None):
         coeff = S.One
         kspec = None
         for h in Mul.make_args(g):
```

Comment:
This could be simplified as
```python
    h_part = defaultdict(list)
    i_part = []
    for g in Add.make_args(f):
        coeff, dep = g.as_coeff_mul(y.func)
        kspec = None
        for h in dep:
             if h.is_Function and h.func == y.func:
                result = h.args[0].match(n + k)
                if result is not None:
                    kspec = int(result[k])
                else:
                    raise ValueError(
                        "'%s(%s + k)' expected, got '%s'" % (y.func, n, h))
            else:
                raise ValueError(
                    "'%s(%s + k)' expected, got '%s'" % (y.func, n, h))
        if kspec is not None:
            h_part[kspec].append(coeff)
        else:
            i_part.append(coeff)
    for k in h_part:
        h_part[k] = Add(*h_part[k])
    i_part = Add(*i_part)
```